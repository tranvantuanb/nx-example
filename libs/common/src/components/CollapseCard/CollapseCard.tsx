import React from 'react';
import {
  Collapse as AntCollapse,
  Divider as AntDivider,
  Space,
  SpaceProps,
} from 'antd';
import tw, { styled } from 'twin.macro';

import { Icon, IconName } from '../Icon';

export interface CollapseCardProps {
  className?: string;
  /**
   * Icon on the left of the header.
   */
  icon?: IconName | JSX.Element;
  /**
   * Header title
   */
  header: React.ReactNode;
  /**
   * Sub header title
   */
  subheader?: React.ReactNode;
  /**
   * Set the collapse open as default
   */
  defaultActive?: boolean;
  /**
   * Collapse card content
   */
  children?: React.ReactNode;
  /**
   * Extra items adding to the right of the header
   */
  extra?: React.ReactNode;
}
export interface CollapseCardEditIconProps {
  className?: string;
  onClick?: (event: React.MouseEvent<HTMLButtonElement>) => void;
}

interface CollapseCardInterface extends React.FC<CollapseCardProps> {
  Divider: typeof Divider;
  EditIcon: typeof EditIcon;
  FooterActions: typeof FooterActions;
}

const StyledCollapseCard = styled(AntCollapse)`
  ${tw`border-gray-200 bg-white`}
  .ant-collapse-item {
    ${tw`border-gray-200`}

    .ant-collapse-arrow {
      transition: all 0.3s ease-out;
      top: 24px !important;
      &.active {
        transform: translate(0, -6px) rotate(-180deg);
      }
    }
  }

  .ant-collapse-content {
    ${tw`border-t-0`}
    .ant-collapse-content-box {
      ${tw`pt-0`}
    }
  }

  form.ant-form {
    ${tw`mx-[8px]`}
  }
`;

const Divider = styled(AntDivider)`
  ${tw`my-[20px] border-gray-200`}
`;

const EditIcon: React.FC<CollapseCardEditIconProps> = ({
  className,
  onClick,
}) => (
  <button
    className={className}
    tw="absolute top-[-12px] right-[0px]"
    onClick={onClick}
  >
    <Icon name="edit" size="16px" />
  </button>
);

const FooterActions: React.FC<SpaceProps> = ({
  children,
  className,
  ...otherSpaceProps
}) => (
  <div className={className} tw="flex flex-row-reverse">
    <Space {...otherSpaceProps}>{children}</Space>
  </div>
);

const Header: React.FC<
  Pick<CollapseCardProps, 'icon' | 'header' | 'subheader'>
> = React.memo(({ icon, header, subheader }) => (
  <div className="header-wrapper flex flex-col">
    <div className="flex flex-row">
      {icon && (
        <div className="header-icon flex items-center mr-[12px]">
          {typeof icon === 'string' ? (
            <Icon name={icon as IconName} size="16px" color="black" />
          ) : (
            icon
          )}
        </div>
      )}
      <div className="header-content font-bold text-base">{header}</div>
    </div>
    {subheader && (
      <div className="subheader-content text-sm text-black-300 mt-[8px]">
        {subheader}
      </div>
    )}
  </div>
));

const Extra: React.FC<Pick<CollapseCardProps, 'extra'>> = React.memo(
  ({ extra }) =>
    !extra ? null : (
      <div
        onClick={(e) => {
          e.preventDefault();
          e.stopPropagation();
        }}
      >
        {extra}
      </div>
    )
);

const CollapseCard: CollapseCardInterface = ({
  className,
  icon,
  header,
  subheader,
  defaultActive = false,
  extra,
  children,
}) => {
  return (
    <StyledCollapseCard
      className={className}
      defaultActiveKey={defaultActive ? ['main-panel'] : undefined}
      expandIcon={({ isActive }) => (
        <Icon name="arrow-down" className={isActive ? 'active' : undefined} />
      )}
      expandIconPosition="end"
    >
      <AntCollapse.Panel
        header={<Header icon={icon} header={header} subheader={subheader} />}
        key="main-panel"
        extra={<Extra extra={extra} />}
      >
        <Divider className="mt-0" />
        <div className="relative">{children}</div>
      </AntCollapse.Panel>
    </StyledCollapseCard>
  );
};

CollapseCard.Divider = Divider;
CollapseCard.EditIcon = EditIcon;
CollapseCard.FooterActions = FooterActions;

export { CollapseCard };
