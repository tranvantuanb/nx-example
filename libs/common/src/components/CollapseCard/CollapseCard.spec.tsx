import { render } from '../../test';

import { CollapseCard } from './CollapseCard';

describe('CollapseCard', () => {
  it('should render successfully', () => {
    const { container } = render(
      <CollapseCard
        header="Header Title"
        subheader="Subheader Content"
        extra={<button>Extra</button>}
        icon="calendar"
      >
        <h1>Card Content</h1>
        <CollapseCard.Divider />
        <h2>Divider</h2>
      </CollapseCard>
    );
    expect(container).toBeTruthy();
    expect(container).toMatchSnapshot();
  });
});
