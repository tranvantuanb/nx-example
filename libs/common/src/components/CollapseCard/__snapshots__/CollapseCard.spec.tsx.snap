// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollapseCard should render successfully 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-right: 12px;
}

.c3 {
  font-weight: 700;
  font-size: 16px;
}

.c4 {
  font-size: 14px;
  line-height: 14px;
  --tw-text-opacity: 1;
  color: rgba(161,167,188,var(--tw-text-opacity));
  margin-top: 8px;
}

<div>
  <div
    class="ant-collapse ant-collapse-icon-position-end css-pkcp1s-StyledCollapseCard eyl3bx71"
  >
    <div
      class="ant-collapse-item"
    >
      <div
        aria-expanded="false"
        class="ant-collapse-header"
        role="button"
        tabindex="0"
      >
        <div
          class="ant-collapse-expand-icon"
        >
          <div
            class="ant-collapse-arrow css-pwgt72-StyledSVGIcon e81aohe0"
          >
            <div>
              <div
                data-src="/icons/arrow-down.svg"
                src=""
              />
            </div>
          </div>
        </div>
        <span
          class="ant-collapse-header-text"
        >
          <div
            class="c0 header-wrapper"
          >
            <div
              class="c1"
            >
              <div
                class="c2 header-icon"
              >
                <div
                  class="css-150499d-StyledSVGIcon-StyledSVGIcon-StyledSVGIcon-StyledSVGIcon e81aohe0"
                  color="black"
                >
                  <div>
                    <div
                      data-src="/icons/calendar.svg"
                      src=""
                    />
                  </div>
                </div>
              </div>
              <div
                class="c3 header-content"
              >
                Header Title
              </div>
            </div>
            <div
              class="c4 subheader-content"
            >
              Subheader Content
            </div>
          </div>
        </span>
        <div
          class="ant-collapse-extra"
        >
          <div>
            <button>
              Extra
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
